### PR TITLE
Creating an MVC structure and handling routes

### PR DESCRIPTION
**Pull request by rkmud**
Creating an MVC structure and handling routes.

Show output: 
Снимок экрана 2024-10-29 в 14 16 56
Снимок экрана 2024-10-29 в 14 17 19

Show output with incorrect routes:
Снимок экрана 2024-10-29 в 14 13 04

### CODE REVIEW

File: app/controllers/ProductController.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace App\Controllers;
+
+class ProductController
+{
+    public function show($id): void
+    {
+
+        $product = $this->getProductById($id);
+
+        if ($product)
```
Comment (vladislavzubetcs) [inline app/controllers/ProductController.php]: Should be wrapped by Response instance, that will handle everything.

ex.:
$this->response->json($data, 200);
$this->response->json(['message' => 'Not Found'], 404);

// future update:
$this->response->view('view.php', $data, $status);

File: index.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+require 'vendor/autoload.php';
+
+$uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
+
+$routes = [
```
Comment (vladislavzubetcs) [inline index.php]: should be moved to the additional php file (configuration file), ex.: routes.php

File: index.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    '/' => ['controller' => 'Home', 'action' => 'index'],
```
Comment (vladislavzubetcs) [inline index.php]: get, post, put and other methods is not supported by these parameters

File: index.php [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    '/about' => ['controller' => 'About', 'action' => 'index'],
+    '/contact' => ['controller' => 'Contact', 'action' => 'index'],
+    '/product/(?P<id>\d+)' => ['controller' => 'Product', 'action' => 'show'],
+];
+
+$routeFlag = false;
```
Comment (vladislavzubetcs) [inline index.php]: all logic that works with the routing should be incapsulated in the Router.php class. If router cannot find route -> response->json(not found)

**Review by vladislavzubetcs — CHANGES_REQUESTED** [review]
well done!

File: app/controllers/ProductController.php (hunk 1 above)
Comment (rkmud) [reply app/controllers/ProductController.php]: Changed

File: app/controllers/AboutController.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+use App\Response\Response;
+
+class AboutController
+{
+    private Response $response;
+
+    public function __construct()
+    {
+        $this->response = new Response();
```
Comment (vladislavzubetcs) [inline app/controllers/AboutController.php]: can be moved to the AbstractController, or BaseController, or simply Controller, and each ur AboutController, ContactController can be extended from it

File: app/controllers/AboutController.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+declare(strict_types=1);
+
+namespace App\Controllers;
+
+use App\Response\Response;
+
+class AboutController
+{
+    private Response $response;
```
Comment (vladislavzubetcs) [inline app/controllers/AboutController.php]: this field also

File: app/controllers/ProductController.php [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+use App\Response\Response;
+
+class ProductController
+{
+    private Response $response;
+
+    public function show($id): void
+    {
+        $product = $this->getProductById($id);
+        $this->response = new Response();
```
Comment (vladislavzubetcs) [inline app/controllers/ProductController.php]: weird, should be in the parent or in the constructor

File: app/controllers/ProductController.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        if ($product) {
```
Comment (vladislavzubetcs) [inline app/controllers/ProductController.php]: https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: app/response/Response.php [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace App\Response;
+
+class Response
+{
+    public function view(string $view, array $data = [], int $status = 200): void
+    {
+        $views_path = sprintf('app/views/%sView.php', $view);
+
+        if (file_exists($views_path))
```
Comment (vladislavzubetcs) [inline app/response/Response.php]: early return also can be applied here

File: app/router/Router.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        $this->uri = parse_url($_SERVER['REQUEST_URI'], PHP_URL_PATH);
+        $this->method = $_SERVER['REQUEST_METHOD'];
+        $this->response = new Response();
+    }
+
+    public function router(): void
+    {
+        foreach ($this->routes as $pattern => $route)
+        {
+            if ($this->method === $route['method'] && preg_match("#^$pattern$#", $this->uri, $matches))
```
Comment (vladislavzubetcs) [inline app/router/Router.php]: early return

File: app/router/Router.php [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                $params = array_filter(
+                    $matches,
+                    function ($key)
```
Comment (vladislavzubetcs) [inline app/router/Router.php]: can be simplified by using arrow function and this function can be static

File: app/views/aboutView.php [hunk 12]
```diff
@@ -0,0 +1,7 @@
+<?php
```
Comment (vladislavzubetcs) [inline app/views/aboutView.php]: these files should be like a HTML templates with the php code inside

File: index.php [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+$params = [];
+
+foreach ($routes as $pattern => $route)
```
Comment (vladislavzubetcs) [inline index.php]: ?

**Review by vladislavzubetcs — CHANGES_REQUESTED** [review]
please, double check PSR code styles and apply it to ur code. I don't want to make 1000 CR comments on ur code style issues :)

File: app/router/Router.php (hunk 10 above)
Comment (rkmud) [reply app/router/Router.php]: changed

**Review by vladislavzubetcs — APPROVED** [review]
good

**Pull request merged**